File: js-bridge/src/protocol.ts
/**
 * Wire protocol for the Pipecat <-> external LLM WebSocket bridge.
 *
 * This is the TypeScript mirror of `src/pipecat_adk/websocket/protocol.py`.
 * Keep the two in sync — they are the two halves of one contract.
 *
 * Messages are UTF-8 JSON text frames. Every message is an envelope with a
 * `v` (protocol version) and a `type`.
 */

export const PROTOCOL_VERSION = 1;

/** Message types. */
export const MSG = {
  // Python -> JS
  SESSION_START: "session.start",
  TURN_RUN: "turn.run",
  TURN_CANCEL: "turn.cancel",
  TURN_COMPLETED: "turn.completed",
  SESSION_END: "session.end",
  PING: "ping",
  // JS -> Python
  SESSION_READY: "session.ready",
  FRAME: "frame",
  TURN_USAGE: "turn.usage",
  STATE_DELTA: "state.delta",
  ERROR: "error",
  PONG: "pong",
} as const;

/** Frame names — the `frame` field of an MSG.FRAME envelope (JS -> Python). */
export const FRAME = {
  RESPONSE_START: "VqlLLMFullResponseStartFrame",
  TEXT: "VqlLLMTextFrame",
  FUNCTION_CALLS_STARTED: "VqlFunctionCallsStartedFrame",
  FUNCTION_CALL_IN_PROGRESS: "VqlFunctionCallInProgressFrame",
  FUNCTION_CALL_RESULT: "VqlFunctionCallResultFrame",
  RESPONSE_END: "VqlLLMFullResponseEndFrame",
} as const;

// ---------------------------------------------------------------------------
// Inbound message shapes (Python -> JS)
// ---------------------------------------------------------------------------

export interface SessionStartMessage {
  v: number;
  type: typeof MSG.SESSION_START;
  session: Record<string, unknown>;
  metadata: Record<string, unknown>;
}

export interface TurnRunMessage {
  v: number;
  type: typeof MSG.TURN_RUN;
  turn_id: string;
  payload: { text?: string } & Record<string, unknown>;
  state_delta: Record<string, unknown> | null;
}

export interface TurnCancelMessage {
  v: number;
  type: typeof MSG.TURN_CANCEL;
  turn_id: string;
}

export interface TurnCompletedMessage {
  v: number;
  type: typeof MSG.TURN_COMPLETED;
  turn_id: string;
  text: string;
  interrupted: boolean;
}

export interface SessionEndMessage {
  v: number;
  type: typeof MSG.SESSION_END;
}

export interface PingMessage {
  v: number;
  type: typeof MSG.PING;
  ts: number;
}

export type InboundMessage =
  | SessionStartMessage
  | TurnRunMessage
  | TurnCancelMessage
  | TurnCompletedMessage
  | SessionEndMessage
  | PingMessage;

// ---------------------------------------------------------------------------
// Outbound payloads (JS -> Python)
// ---------------------------------------------------------------------------

export interface FunctionCall {
  tool_call_id: string;
  function_name: string;
  arguments: Record<string, unknown>;
}

export interface TokenUsage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  cache_read_input_tokens: number;
  reasoning_tokens: number;
}

export class ProtocolError extends Error {}

/** Parse and validate a raw WebSocket message into an inbound envelope. */
export function decodeMessage(raw: string | Uint8Array): InboundMessage {
  const text =
    typeof raw === "string" ? raw : new TextDecoder().decode(raw);

  let envelope: unknown;
  try {
    envelope = JSON.parse(text);
  } catch (e) {
    throw new ProtocolError(`message is not valid JSON: ${(e as Error).message}`);
  }
  if (typeof envelope !== "object" || envelope === null) {
    throw new ProtocolError("message must be a JSON object");
  }
  const env = envelope as Record<string, unknown>;
  if (env.v !== PROTOCOL_VERSION) {
    throw new ProtocolError(
      `unsupported protocol version ${String(env.v)} (expected ${PROTOCOL_VERSION})`,
    );
  }
  if (typeof env.type !== "string" || env.type.length === 0) {
    throw new ProtocolError("message is missing 'type'");
  }
  return env as unknown as InboundMessage;
}

// ---------------------------------------------------------------------------
// Encoders (JS -> Python)
// ---------------------------------------------------------------------------

function encode(obj: Record<string, unknown>): string {
  return JSON.stringify({ v: PROTOCOL_VERSION, ...obj });
}

export function encodeSessionReady(info: Record<string, unknown> = {}): string {
  return encode({ type: MSG.SESSION_READY, info });
}

export function encodePong(ts: number): string {
  return encode({ type: MSG.PONG, ts });
}

function encodeFrame(
  frame: string,
  turnId: string,
  data: Record<string, unknown>,
): string {
  return encode({ type: MSG.FRAME, frame, turn_id: turnId, data });
}

export function encodeResponseStart(turnId: string, invocationId: string): string {
  return encodeFrame(FRAME.RESPONSE_START, turnId, { invocation_id: invocationId });
}

export function encodeResponseEnd(turnId: string, invocationId: string): string {
  return encodeFrame(FRAME.RESPONSE_END, turnId, { invocation_id: invocationId });
}

export function encodeText(
  turnId: string,
  invocationId: string,
  text: string,
): string {
  return encodeFrame(FRAME.TEXT, turnId, { invocation_id: invocationId, text });
}

export function encodeFunctionCallsStarted(
  turnId: string,
  invocationId: string,
  calls: FunctionCall[],
): string {
  return encodeFrame(FRAME.FUNCTION_CALLS_STARTED, turnId, {
    invocation_id: invocationId,
    function_calls: calls,
  });
}

export function encodeFunctionCallInProgress(
  turnId: string,
  invocationId: string,
  call: FunctionCall,
): string {
  return encodeFrame(FRAME.FUNCTION_CALL_IN_PROGRESS, turnId, {
    invocation_id: invocationId,
    tool_call_id: call.tool_call_id,
    function_name: call.function_name,
    arguments: call.arguments,
  });
}

export function encodeFunctionCallResult(
  turnId: string,
  invocationId: string,
  call: FunctionCall,
  result: unknown,
): string {
  return encodeFrame(FRAME.FUNCTION_CALL_RESULT, turnId, {
    invocation_id: invocationId,
    tool_call_id: call.tool_call_id,
    function_name: call.function_name,
    result,
  });
}

export function encodeTurnUsage(
  turnId: string,
  usage: Partial<TokenUsage>,
): string {
  return encode({ type: MSG.TURN_USAGE, turn_id: turnId, data: usage });
}

export function encodeStateDelta(
  turnId: string,
  delta: Record<string, unknown>,
): string {
  return encode({ type: MSG.STATE_DELTA, turn_id: turnId, data: delta });
}

export function encodeError(
  message: string,
  opts: { turnId?: string; fatal?: boolean } = {},
): string {
  return encode({
    type: MSG.ERROR,
    turn_id: opts.turnId ?? null,
    data: { message, fatal: opts.fatal ?? false },
  });
}

File: src/pipecat_adk/websocket/protocol.py
"""Wire protocol for the Pipecat <-> external LLM WebSocket bridge.

This module is the single source of truth for the bridge protocol.  The
TypeScript implementation in ``js-bridge/src/protocol.ts`` mirrors these
constants and shapes; keep the two in sync.

Transport
---------
Messages are UTF-8 JSON text frames over a WebSocket connection.  Every
message is an envelope with a ``v`` (protocol version) and a ``type`` field.

Direction
---------
- Python -> JS: ``session.start``, ``turn.run``, ``turn.cancel``,
  ``turn.completed``, ``session.end``, ``ping``.
- JS -> Python: ``session.ready``, ``frame``, ``turn.usage``, ``state.delta``,
  ``error``, ``pong``.

A turn
------
``WebSocketLLMService`` sends one ``turn.run`` per ``VqlContextFrame``.  The
JS component answers with a stream of ``frame`` messages that always begins
with ``VqlLLMFullResponseStartFrame`` and ends with
``VqlLLMFullResponseEndFrame`` — the end frame is the stream terminator.
``turn.cancel`` aborts an in-flight turn (pipeline interruption).
"""

from __future__ import annotations

import json
from typing import Any, Optional

from pipecat.frames.frames import (
    ErrorFrame,
    Frame,
    FunctionCallFromLLM,
)
from pipecat.processors.frame_processor import FrameDirection

from ..frames import (
    VqlFunctionCallInProgressFrame,
    VqlFunctionCallResultFrame,
    VqlFunctionCallsStartedFrame,
    VqlLLMFullResponseEndFrame,
    VqlLLMFullResponseStartFrame,
    VqlLLMTextFrame,
)

PROTOCOL_VERSION = 1

# ---------------------------------------------------------------------------
# Message types
# ---------------------------------------------------------------------------

# Python -> JS
MSG_SESSION_START = "session.start"
MSG_TURN_RUN = "turn.run"
MSG_TURN_CANCEL = "turn.cancel"
MSG_TURN_COMPLETED = "turn.completed"
MSG_SESSION_END = "session.end"
MSG_PING = "ping"

# JS -> Python
MSG_SESSION_READY = "session.ready"
MSG_FRAME = "frame"
MSG_TURN_USAGE = "turn.usage"
MSG_STATE_DELTA = "state.delta"
MSG_ERROR = "error"
MSG_PONG = "pong"

# ---------------------------------------------------------------------------
# Frame names — the ``frame`` field of an MSG_FRAME envelope (JS -> Python)
# ---------------------------------------------------------------------------

FRAME_RESPONSE_START = "VqlLLMFullResponseStartFrame"
FRAME_TEXT = "VqlLLMTextFrame"
FRAME_FUNCTION_CALLS_STARTED = "VqlFunctionCallsStartedFrame"
FRAME_FUNCTION_CALL_IN_PROGRESS = "VqlFunctionCallInProgressFrame"
FRAME_FUNCTION_CALL_RESULT = "VqlFunctionCallResultFrame"
FRAME_RESPONSE_END = "VqlLLMFullResponseEndFrame"

# Pipeline direction(s) each output frame is pushed in.  Function-call frames
# go both ways (upstream so STTMuteFilter can mute the mic, downstream so the
# UI can render a "thinking" state) — this mirrors AdkLLMService exactly.
_BOTH = (FrameDirection.UPSTREAM, FrameDirection.DOWNSTREAM)
_DOWN = (FrameDirection.DOWNSTREAM,)

FRAME_DIRECTIONS: dict[str, tuple[FrameDirection, ...]] = {
    FRAME_RESPONSE_START: _DOWN,
    FRAME_TEXT: _DOWN,
    FRAME_FUNCTION_CALLS_STARTED: _BOTH,
    FRAME_FUNCTION_CALL_IN_PROGRESS: _BOTH,
    FRAME_FUNCTION_CALL_RESULT: _BOTH,
    FRAME_RESPONSE_END: _DOWN,
}


class ProtocolError(Exception):
    """Raised when a message cannot be parsed or violates the protocol."""


# ---------------------------------------------------------------------------
# Encoding — Python -> JS
# ---------------------------------------------------------------------------


def _dumps(obj: dict[str, Any]) -> str:
    return json.dumps(obj, separators=(",", ":"), ensure_ascii=False)


def encode_session_start(
    session: Optional[dict[str, Any]] = None,
    metadata: Optional[dict[str, Any]] = None,
) -> str:
    """Handshake sent by the Python side as soon as the socket opens."""
    return _dumps(
        {
            "v": PROTOCOL_VERSION,
            "type": MSG_SESSION_START,
            "session": session or {},
            "metadata": metadata or {},
        }
    )


def encode_turn_run(
    turn_id: str,
    payload: dict[str, Any],
    state_delta: Optional[dict[str, Any]] = None,
) -> str:
    """Request the JS component to generate a response for ``turn_id``."""
    return _dumps(
        {
            "v": PROTOCOL_VERSION,
            "type": MSG_TURN_RUN,
            "turn_id": turn_id,
            "payload": payload,
            "state_delta": state_delta,
        }
    )


def encode_turn_cancel(turn_id: str) -> str:
    """Abort an in-flight turn (the pipeline was interrupted)."""
    return _dumps({"v": PROTOCOL_VERSION, "type": MSG_TURN_CANCEL, "turn_id": turn_id})


def encode_turn_completed(turn_id: str, text: str, interrupted: bool) -> str:
    """Report what the user actually heard once a turn finishes.

    The JS component uses this to write its own ``[HEARD]`` provenance, the
    same way AdkLLMService writes a ``[HEARD]`` event into the ADK session.
    """
    return _dumps(
        {
            "v": PROTOCOL_VERSION,
            "type": MSG_TURN_COMPLETED,
            "turn_id": turn_id,
            "text": text,
            "interrupted": interrupted,
        }
    )


def encode_session_end() -> str:
    """Signal a graceful shutdown before closing the socket."""
    return _dumps({"v": PROTOCOL_VERSION, "type": MSG_SESSION_END})


def encode_ping(ts: float) -> str:
    """Application-level heartbeat — detects a hung JS event loop."""
    return _dumps({"v": PROTOCOL_VERSION, "type": MSG_PING, "ts": ts})


# ---------------------------------------------------------------------------
# Decoding — JS -> Python
# ---------------------------------------------------------------------------


def decode_message(raw: str | bytes) -> dict[str, Any]:
    """Parse and validate a raw WebSocket message into an envelope dict.

    Raises:
        ProtocolError: if the payload is not JSON, not an object, has the
            wrong protocol version, or is missing a ``type``.
    """
    if isinstance(raw, bytes):
        try:
            raw = raw.decode("utf-8")
        except UnicodeDecodeError as e:
            raise ProtocolError(f"message is not valid UTF-8: {e}") from e

    try:
        envelope = json.loads(raw)
    except json.JSONDecodeError as e:
        raise ProtocolError(f"message is not valid JSON: {e}") from e

    if not isinstance(envelope, dict):
        raise ProtocolError(f"message must be a JSON object, got {type(envelope).__name__}")

    version = envelope.get("v")
    if version != PROTOCOL_VERSION:
        raise ProtocolError(
            f"unsupported protocol version {version!r} (expected {PROTOCOL_VERSION})"
        )

    if not envelope.get("type"):
        raise ProtocolError("message is missing 'type'")

    return envelope


def build_output_frame(envelope: dict[str, Any]) -> tuple[Frame, tuple[FrameDirection, ...]]:
    """Reconstruct a Vql frame (and its push direction) from an MSG_FRAME envelope.

    Returns:
        A ``(frame, directions)`` pair.  ``directions`` is the tuple of
        FrameDirection values the frame must be pushed in.

    Raises:
        ProtocolError: if the frame name is unknown or required fields are missing.
    """
    name = envelope.get("frame")
    if name not in FRAME_DIRECTIONS:
        raise ProtocolError(f"unknown output frame {name!r}")

    turn_id = envelope.get("turn_id")
    if not turn_id:
        raise ProtocolError(f"frame {name!r} is missing 'turn_id'")

    data = envelope.get("data") or {}
    invocation_id = data.get("invocation_id", "")
    directions = FRAME_DIRECTIONS[name]

    if name == FRAME_RESPONSE_START:
        return (
            VqlLLMFullResponseStartFrame(turn_id=turn_id, invocation_id=invocation_id),
            directions,
        )

    if name == FRAME_RESPONSE_END:
        return (
            VqlLLMFullResponseEndFrame(turn_id=turn_id, invocation_id=invocation_id),
            directions,
        )

    if name == FRAME_TEXT:
        text = data.get("text")
        if text is None:
            raise ProtocolError("VqlLLMTextFrame is missing 'text'")
        return (
            VqlLLMTextFrame(text=text, turn_id=turn_id, invocation_id=invocation_id),
            directions,
        )

    if name == FRAME_FUNCTION_CALLS_STARTED:
        raw_calls = data.get("function_calls") or []
        calls = [_build_function_call_from_llm(c) for c in raw_calls]
        return (
            VqlFunctionCallsStartedFrame(
                function_calls=calls, turn_id=turn_id, invocation_id=invocation_id
            ),
            directions,
        )

    if name == FRAME_FUNCTION_CALL_IN_PROGRESS:
        tool_call_id, function_name = _require_call_ids(data, name)
        return (
            VqlFunctionCallInProgressFrame(
                function_name=function_name,
                tool_call_id=tool_call_id,
                arguments=data.get("arguments"),
                turn_id=turn_id,
                invocation_id=invocation_id,
            ),
            directions,
        )

    # FRAME_FUNCTION_CALL_RESULT
    tool_call_id, function_name = _require_call_ids(data, name)
    return (
        VqlFunctionCallResultFrame(
            function_name=function_name,
            tool_call_id=tool_call_id,
            arguments=None,
            result=data.get("result") or {},
            turn_id=turn_id,
            invocation_id=invocation_id,
        ),
        directions,
    )


def build_error_frame(envelope: dict[str, Any]) -> ErrorFrame:
    """Build an ErrorFrame from an MSG_ERROR envelope."""
    data = envelope.get("data") or {}
    message = data.get("message") or "remote LLM bridge error"
    return ErrorFrame(error=str(message), fatal=bool(data.get("fatal", False)))


def _require_call_ids(data: dict[str, Any], frame_name: str) -> tuple[str, str]:
    tool_call_id = data.get("tool_call_id")
    function_name = data.get("function_name")
    if not tool_call_id or not function_name:
        raise ProtocolError(
            f"{frame_name} requires 'tool_call_id' and 'function_name'"
        )
    return tool_call_id, function_name


def _build_function_call_from_llm(call: dict[str, Any]) -> FunctionCallFromLLM:
    tool_call_id = call.get("tool_call_id")
    function_name = call.get("function_name")
    if not tool_call_id or not function_name:
        raise ProtocolError(
            "function_calls entry requires 'tool_call_id' and 'function_name'"
        )
    return FunctionCallFromLLM(
        tool_call_id=tool_call_id,
        function_name=function_name,
        arguments=call.get("arguments") or {},
        context=None,
    )
